### PR TITLE
[RISCV] Merge GPRPair and GPRF64Pair

### PR DESCRIPTION
**Pull request by lenary**
As suggested by Craig, this tries to merge the two sets of register classes created in #112983, GPRPair* and GPRF64Pair*.

- I added some explicit annotations to `RISCVInstrInfoD.td` which fixed the type inference issues I was seeing from tablegen for select patterns.
- I've had to make the behaviour of `splitValueIntoRegisterParts` and `joinRegisterPartsIntoValue` cover more cases, because you cannot bitcast to/from untyped (the bitcast would otherwise have been inserted automatically by TargetLowering code).
- I apparently didn't need to change `getNumRegisters` again, which continues to tell me there's a bug in the code for tied inputs. I added some more test coverage of this case but it didn't seem to help find the asserts I was finding before - I think the difference is between the default behaviour for integers which doesn't apply to floats.
- There's still a difference between BuildGPRPair and BuildPairF64 (and the same for SplitGPRPair and SplitF64). I'm not happy with this, I think it's quite confusing, as they're very similar, just differing in whether they give a `untyped` or a `f64`. I haven't really worked out how the DAGCombiner copes if one meets the other, I know we have some of this for the f64 variants already, but they're a lot more complex than the GPRPair variants anyway.

### CODE REVIEW

**Comment by lenary** [comment]
@topperc what do you think about this approach? Do you think I should be working harder to merge `BuildGPRPair` and `BuildPairF64` (and similarly for Split)

File: llvm/lib/Target/RISCV/RISCVISelDAGToDAG.cpp [hunk 1]
```diff
@@ -982,13 +976,15 @@ void RISCVDAGToDAGISel::Select(SDNode *Node) {
              "SplitF64 only handled here on rv32i_zdinx");
 
       if (!SDValue(Node, 0).use_empty()) {
-        SDValue Lo = CurDAG->getTargetExtractSubreg(RISCV::sub_gpr_even, DL, VT,
+        SDValue Lo = CurDAG->getTargetExtractSubreg(RISCV::sub_gpr_even, DL,
+                                                    Node->getSimpleValueType(0),
```
Comment (topperc) [inline llvm/lib/Target/RISCV/RISCVISelDAGToDAG.cpp:980]: getValueType(0)?

**Comment by topperc** [comment]
> @topperc what do you think about this approach? Do you think I should be working harder to merge `BuildGPRPair` and `BuildPairF64` (and similarly for Split)

I'm ok with them being separate. It gives some degree of type safety.

File: llvm/lib/Target/RISCV/RISCVISelDAGToDAG.cpp (hunk 1 above)
Comment (lenary) [reply llvm/lib/Target/RISCV/RISCVISelDAGToDAG.cpp]: Addressed (and in the other call)

**Comment by lenary** [comment]
> > @topperc what do you think about this approach? Do you think I should be working harder to merge `BuildGPRPair` and `BuildPairF64` (and similarly for Split)
> 
> I'm ok with them being separate. It gives some degree of type safety.

Cool.

I'm not sure there's much more to do, but maybe I missed something?

**Review by topperc — APPROVED** [review]
LGTM

**Pull request merged**